src/cocotb_tools/pytest/controller.py

```diff
@@ -0,0 +1,307 @@
+        # Otherwise pytest can raise an exception when collecting test items
+        #
+        # Fun fact, this will not effect collection of test items done by pytest sub-process (simulator),
+        # because collection of test items is done independently from it and with valid cocotb.top
+        #
+        # This may effect collected items when invoking pytest --collect-only
+        # Some tests could be wrongly deselected when using boolean condition like this
+        # @pytest.mark.skipif(cocotb.top.PARAMETER == 1) but this harmless
+        #
+        # TODO: Get version of HDL simulator, mostly: <command> --version
+        # TODO: Get names and values of HDL parameters/generics, parse HDL source files?
+        setattr(cocotb, "SIM_NAME", get_simulator(config))
+        setattr(cocotb, "SIM_VERSION", "")
+        setattr(cocotb, "top", MockSimHandle())
+
+    @hookimpl(tryfirst=True, wrapper=True)
+    def pytest_pycollect_makeitem(
+        self, collector: Module | Class, name: str, obj: object
+    ) -> Generator[Item | Collector | list[Item | Collector] | None, None, None]:
+        """Collect cocotb runners and cocotb tests from Python modules.
+
+        Args:
+            collector: Python module or class containing cocotb runners or cocotb tests.
+            name: Name of collected item (test function, cocotb runner, cocotb test).
+            obj: Object representation of collected item (test function, cocotb runner, cocotb test).
+
+        Yields:
+            Collected test function, cocotb runner or cocotb test.
+        """
+        result: Item | Collector | list[Item | Collector] | None = yield
+
+        if result is None:
+            return None
+
+        items: Iterable[Item | Collector] = (
+            result if isinstance(result, list) else (result,)
+        )
+
+        return list(self._collect(collector, items))
+
+    @hookimpl(tryfirst=True)
+    def pytest_runtest_setup(self, item: Item) -> None:
+        """Setup environment for pytest sub-process (simulator)
+        that will be started as test function from the main pytest parent process.
+
+        Args:
+            item: Test function.
+        """
+        # Expose cocotb runner nodeid and keywords from the main pytest parent process
+        # to pytest sub-process (simulator) via environment variables
+        os.environ["COCOTB_PYTEST_NODEID"] = item.nodeid
+        os.environ["COCOTB_PYTEST_KEYWORDS"] = ",".join(item.keywords)
+
+    @staticmethod
+    def _split_nodeid(nodeid: str) -> tuple[PurePosixPath, str]:
+        """Split provided node identifier to path and function name.
+
+        Args:
+            nodeid: Item nodeid in form of ``<path_to_file>::[<class_name>::]<function_name>``.
+
+        Returns:
+            Two-elements tuple with path to file and name of function with scope.
+        """
+        (path, _, function) = nodeid.partition("::")
+
+        return PurePosixPath(path), function
```

Comment:
Instantiating pure paths is a code smell.
```suggestion
        return PurePosixPath(path), function
```

Reply:
@ktbarrett Removed in this commit db63a4d7662502b4f4b9366e53de4471f3d3f6f9